### PR TITLE
Give built-in macros stable addresses in the standard library

### PR DESCRIPTION
**Pull request by petrochenkov**
Continuation of https://github.com/rust-lang/rust/pull/62086.

Derive macros corresponding to traits from libcore are now available through the same paths as those traits:
- `Clone` - `{core,std}::clone::Clone`
- `PartialEq` - `{core,std}::cmp::PartialEq`
- `Eq` - `{core,std}::cmp::Eq`
- `PartialOrd` - `{core,std}::cmp::PartialOrd`
- `Ord` - `{core,std}::cmp::Ord`
- `Default` - `{core,std}::default::Default`
- `Debug` - `{core,std}::fmt::Debug`
- `Hash` - `{core,std}::hash::Hash`
- `Copy` - `{core,std}::marker::Copy`

Fn-like built-in macros are now available through libcore and libstd's root module, by analogy with non-builtin macros defined by libcore and libstd:
```rust
{core,std}::{
    __rust_unstable_column,
    asm,
    assert,
    cfg,
    column,
    compile_error,
    concat,
    concat_idents,
    env,
    file,
    format_args,
    format_args_nl,
    global_asm,
    include,
    include_bytes,
    include_str,
    line,
    log_syntax,
    module_path,
    option_env,
    stringify,
    trace_macros,
}
```

Derive macros without a corresponding trait in libcore or libstd are still available only through prelude (also see https://github.com/rust-lang/rust/pull/62507).
Attribute macros also keep being available only through prelude, mostly because they don't have an existing practice to follow. An advice from the library team on their eventual placement would be appreciated.
```rust
    RustcDecodable,
    RustcEncodable,
    bench,
    global_allocator,
    test,
    test_case,
```

r? @alexcrichton

### CODE REVIEW

**Review by alexcrichton — COMMENTED** [review]
As for my opinions about the other attributes, I think in general this is a good PR and wouldn't want to bog it down too much with where to put the remaining macros. They're pretty inconsequential so I think it's fine to deal with them in follow-up PRs. That being said some thoughts I'd have are:

* `Encodable`/`Decodable` - being removed soon in https://github.com/rust-lang/rust/pull/62507 (yay!) so probably not too relevant
* `RustcEncodable`/`RustcDecodable` - I think it's fine to relegate these to "only in the prelude" for now, there's basically zero new users of these macros so it's basically a game of "what's the minimum we can do to keep what's working still working"
* `global_allocator` - my guess would be `core::alloc` as a good place for this. 
* `test`, `test_case`, `bench` - my hunch is that as we stabilize `#[test_case]` we're going to create a `std::test` module. In that world they'd naturally all fit into `std::test`. For now though I think leaving them in the prelude is fine.

**Comment by alexcrichton** [comment]
I'm tweaking the tags since I'll run this through rfcbot once some of the review above has been addressed.

File: src/libcore/prelude/v1.rs [hunk 1]
```diff
@@ -51,17 +51,10 @@ pub use crate::result::Result::{self, Ok, Err};
...
-    Default,
     Encodable,
-    Eq,
     Hash,
-    Ord,
-    PartialEq,
-    PartialOrd,
     RustcDecodable,
     RustcEncodable,
     __rust_unstable_column,
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply src/libcore/prelude/v1.rs]: Yes, the macros are not available with `#![no_core]`, you have to define them yourself, similarly to lang items.
Some `no_core` tests in the test suite do that.

File: src/libcore/clone.rs [hunk 2]
```diff
@@ -133,6 +133,12 @@ pub trait Clone : Sized {
     }
 }
 
+/// Derive macro generating an impl of the trait `Clone`.
+#[cfg(not(bootstrap))]
+#[stable(feature = "builtin_macro_prelude", since = "1.38.0")]
+#[doc(inline)]
+pub use crate::macros::builtin::Clone;
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply src/libcore/clone.rs]: That's what I originally tried, but encountered some complications.

For example, `core::prelude::v1` wants to reexport the macro `Debug` from `core::fmt`, but not the trait `Debug`.

So, at least *some* of the macros need to be defined in dedicated modules to be able to be `use`d separately from other entities with the same name, so I kept them all defined in the dedicated module `core::macros::builtin`.
(IIRC, there were also some problems with rustdoc, not sure about this.)

I can try to rearrange them and create separate modules only when necessary, e.g.
```rust
pub mod fmt {
    pub trait Debug { ... }

    pub(crate) mod macros {
        macro Debug { ... }
    }
    pub use macros::Debug;
}

pub mod prelude::v1 {
    pub use crate::fmt::macros::
}
```
and see what happens.

If we ever grow an ability to import in a single namespace (like `use foo::bar in macro`), then these helper modules could be removed.

[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/libcore/clone.rs]: Ah ok, thanks for the clarification! I think it's fine to leave as-is in that case. I'll add this to the fcp comment and see if others have a preference. I would have a mild preference for moving all macros that can to be inline in the module itself and using a workaround like you mentioned for `Debug` and others which need similar treatment.

If there's other weirder rustdoc bugs though it's fine to leave this as is and we can deal with it at another time, it's just a stylistic issue.

**Comment by petrochenkov** [comment]
Updated.
- Derive macros are moved into modules to avoid reexports.
- `pub macro` is converted to `#[macro_export] macro_rules` for fn-like macros to avoid more reexports.
- Other comments are addressed.

File: src/libcore/lib.rs [hunk 3]
```diff
@@ -243,3 +243,31 @@ mod core_arch;
 
 #[stable(feature = "simd_arch", since = "1.27.0")]
 pub use core_arch::arch;
+
+#[cfg(not(bootstrap))]
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply src/libcore/lib.rs]: No longer relevant after the conversion to `#[macro_export] macro_rules`.

File: src/libstd/lib.rs [hunk 4]
```diff
@@ -515,31 +510,52 @@ mod std_detect;
...
+    debug_assert_ne,
+    debug_assert,
+    r#try,
+    unimplemented,
+    unreachable,
+    write,
+    writeln,
+    // Unstable
+    todo,
+    uninit_array,
```
Comment (petrochenkov) [inline src/libstd/lib.rs]: Drive-by fix - recently introduced `uninit_array` was defined in core, but not reexported in std.

**Comment by alexcrichton** [comment]
@rfcbot fcp merge

Ok this all looks great to me, thanks again @petrochenkov!

As a heads up to the @rust-lang/libs folks, this is a follow-up to https://github.com/rust-lang/rust/pull/62086 where builtin macros now are earning stable locations throughout the standard library (e.g. libcore and libstd). The diff is pretty easy to read and the description is comprehensive as well. Want to get a sign off though since this will actually affect some stable surface area!

**Comment by SimonSapin** [comment]
When defining a proc macro, the choice of using `#[proc_macro]` v.s. `#[proc_macro_derive(…)]` v.s. `#[proc_macro_attribute]` determines what kind of macro this is.

What is the equivalent of that for `pub macro` items? For example, what in this PR makes `#[derive(Clone)]` valid but not `Clone!(…)` ?

**Comment by petrochenkov** [comment]
@SimonSapin 
The compiler maintains a pre-defined map `Map<Name, Signature>` of built-in macros.
When the compiler sees a "lang item" macro (`#[rustc_builtin_macro]`) , it looks it signature up in that map. The macro kind (bang/attr/derive) is a [part of that signature](https://github.com/rust-lang/rust/blob/f690098e6d65ad7b33dc7fdefccc387806782027/src/libsyntax/ext/base.rs#L521-L571).

This machinery was implemented earlier in https://github.com/rust-lang/rust/pull/62086 rather than in this PR.

**Comment by SimonSapin** [comment]
If the signature is built into the compiler, does that mean the `$item:item` part of `pub macro Clone($item:item) { /* compiler built-in */ }` is ignored? (Except perhaps by rustdoc?)

**Comment by petrochenkov** [comment]
Yes, the `$item:item` is for rustdoc.

**Comment by SimonSapin** [comment]
Sounds good, thanks!

File: src/libstd/lib.rs (hunk 4 above)
Comment (RalfJung) [reply src/libstd/lib.rs]: FWIW, `uninit_array` should go away with the next bootstrap bump anyway.

Comment (Centril) [reply src/libstd/lib.rs]: You can remove the macro if possible; it is rendered obsolete by https://github.com/rust-lang/rust/issues/49147.

Comment (petrochenkov) [reply src/libstd/lib.rs]: Ah, ok.
Will remove.

Comment (RalfJung) [reply src/libstd/lib.rs]: You cannot remove it yet because https://github.com/rust-lang/rust/issues/49147 isn't in the bootstrap compiler yet.

Comment (petrochenkov) [reply src/libstd/lib.rs]: The reexport, I mean.

**Comment by alexcrichton** [comment]
r=me with a rebase
